pkg/kubectl/cmd/run.go

```diff
@@ -372,81 +374,75 @@ func contains(resourcesList map[string]*unversioned.APIResourceList, resource un
 
 // waitForPod watches the given pod until the exitCondition is true. Each two seconds
 // the tick function is called e.g. for progress output.
-func waitForPod(c *client.Client, ns, name string, exitCondition func(*api.Pod) bool, tick func(*api.Pod)) (*api.Pod, error) {
-	pod, err := c.Pods(ns).Get(name)
+func waitForPod(c *client.Client, ns, name string, exitCondition watch.ConditionFunc, tick func(*api.Pod)) (*api.Pod, error) {
+	w, err := c.Pods(ns).Watch(api.SingleObject(api.ObjectMeta{Name: name}))
 	if err != nil {
 		return nil, err
 	}
-	if exitCondition(pod) {
-		return pod, nil
-	}
 
-	tick(pod)
+	pods := make(chan *api.Pod) // observed pods passed to the exitCondition
+	defer close(pods)
 
-	w, err := c.Pods(ns).Watch(api.SingleObject(api.ObjectMeta{Name: pod.Name, ResourceVersion: pod.ResourceVersion}))
-	if err != nil {
-		return nil, err
-	}
-
-	t := time.NewTicker(2 * time.Second)
-	defer t.Stop()
+	// wait for the first event, then start the 2 sec ticker and loop
 	go func() {
-		for range t.C {
-			tick(pod)
+		pod := <-pods
+		if pod == nil {
+			return
 		}
-	}()
+		tick(pod)
+
+		t := time.NewTicker(2 * time.Second)
+		defer t.Stop()
 
-	err = nil
-	result := pod
-	kubectl.WatchLoop(w, func(ev watch.Event) error {
-		switch ev.Type {
-		case watch.Added, watch.Modified:
-			pod = ev.Object.(*api.Pod)
-			if exitCondition(pod) {
-				result = pod
-				w.Stop()
+		for {
+			select {
+			case pod = <-pods:
+				if pod == nil {
+					return
+				}
+			case _, ok := <-t.C:
+				if !ok {
+					return
+				}
+				tick(pod)
 			}
-		case watch.Deleted:
-			w.Stop()
-		case watch.Error:
-			result = nil
-			err = fmt.Errorf("failed to watch pod %s/%s", ns, name)
-			w.Stop()
 		}
-		return nil
-	})
+	}()
 
+	intr := interrupt.New(nil, w.Stop)
```

Comment:
we could pass in a interrupt context and `Chain` it with `w.Stop`. But `waitForPod` is private here. So I keep it simple.